src/async.ts

```diff
@@ -126,60 +156,52 @@ function asyncMemoizer<T1, T2, T3, T4, T5, T6, TResult>(
       key = hash(...parameters);
     }
 
-    var fromCache = cache.get(key);
-
+    const fromCache = cache.get(key);
     if (fromCache) {
       emit('hit', ...parameters);
+      // found, invoke callback
+      return runCallbacks([callback], [null].concat(fromCache));
+    }
 
-      if (clone) {
-        return callback(...[null].concat(fromCache).map(cloneDeep));
-      }
-      return callback(...[null].concat(fromCache));
+    const pendingLoad = loading.get(key);
+    if (pendingLoad && pendingLoad.expiresAt > Date.now()) {
+      // request already in progress, queue and return
+      pendingLoad.queue.push(callback);
+      emit('queue', ...parameters);
+      return;
     }
 
-    if (!loading.get(key)) {
-      emit('miss', ...parameters);
+    emit('miss', ...parameters);
+
+    // no pending request or not resolved before expiration
+    // create a new queue and invoke load
+    const queue = [ callback ];
+    loading.set(key, {
+      queue,
+      expiresAt: Date.now() + queueMaxAge
+    });
+
+    const started = Date.now();
```

Comment:
Why call `Date.now` twice (and potentially have a minor diff)?

Reply:
Good catch, I switched to expiresAt at a second iteration and didn't notice 👌 . Changed.